airbyte-integrations/connectors/source-slack/source_slack/components/join_channels.py

```diff
@@ -0,0 +1,122 @@
+        return []
+
+    def request_body_json(self, stream_slice: Mapping = None, **kwargs) -> Optional[Mapping]:
+        return {"channel": stream_slice["channel"]}
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        """
+        The pagination is not applicable to this Service Stream.
+        """
+        return None
+
+
+class ChannelsRetriever(SimpleRetriever):
+    def __post_init__(self, parameters: Mapping[str, Any]):
+        super().__post_init__(parameters)
+        self.stream_slicer = SinglePartitionRouter(parameters={})
+        self.record_selector.transformations = []
+
+    def should_join_to_channel(self, config: Mapping[str, Any], record: Record) -> bool:
+        """
+        The `is_member` property indicates whether the API Bot is already assigned / joined to the channel.
+        https://api.slack.com/types/conversation#booleans
+        """
+        return config["join_channels"] and not record.get("is_member")
+
+    def make_join_channel_slice(self, channel: Mapping[str, Any]) -> Mapping[str, Any]:
+        channel_id: str = channel.get("id")
+        channel_name: str = channel.get("name")
+        LOGGER.info(f"Joining Slack Channel: `{channel_name}`")
+        return {"channel": channel_id, "channel_name": channel_name}
+
+    def join_channels_stream(self, config) -> JoinChannelsStream:
+        token = config["credentials"].get("api_token") or config["credentials"].get("access_token")
+        authenticator = TokenAuthenticator(token)
+        channel_filter = config["channel_filter"]
+        return JoinChannelsStream(authenticator=authenticator, channel_filter=channel_filter)
+
+    def join_channel(self, config: Mapping[str, Any], record: Mapping[str, Any]):
+        list(
+            self.join_channels_stream(config).read_records(
+                sync_mode=SyncMode.full_refresh,
+                stream_slice=self.make_join_channel_slice(record),
+            )
+        )
+
+    def read_records(
+        self,
+        records_schema: Mapping[str, Any],
+        stream_slice: Optional[StreamSlice] = None,
+    ) -> Iterable[StreamData]:
+        _slice = stream_slice or StreamSlice(partition={}, cursor_slice={})  # None-check
+
+        self._paginator.reset()
+
+        most_recent_record_from_slice = None
+        record_generator = partial(
+            self._parse_records,
+            stream_state=self.state or {},
+            stream_slice=_slice,
+            records_schema=records_schema,
+        )
+
+        for stream_data in self._read_pages(record_generator, self.state, _slice):
+            current_record = self._extract_record(stream_data, _slice)
+            if self.cursor and current_record:
+                self.cursor.observe(_slice, current_record)
+
+            most_recent_record_from_slice = self._get_most_recent_record(most_recent_record_from_slice, current_record, _slice)
+            # joining channel logic
```

Comment:
This check should be the first one in the loop, after we have the `stream_data`, Once the channel data has been received as a `stream_data`, we then check whether or not we are joined already (is_member == True), if so - perform all other steps (STATE management) with the records inside the `stream_data` and yield, otherwise - join. 

Basically, can we move the line `115`, to `109` ?

Reply:
moved to 110